cmd/queue/execprobe.go

```diff
@@ -116,9 +118,15 @@ func probeQueueHealthPath(timeout time.Duration, queueServingPort int) error {
 			// Return nil error for retrying
 			return false, nil
 		}
-		defer res.Body.Close()
 
-		// fail readiness immediately rather than retrying if we get a header indicating we're shutting down.
+		defer func() {
+			// Ensure body is read and closed to ensure connection can be re-used via keep-alive.
+			// No point handling errors here, connection just won't be reused.
+			io.Copy(ioutil.Discard, res.Body)
```

Comment:
Should we add the same to `HTTPProbe` so connections further upstream are properly kept alive too?

Reply:
yeah, this case happens a little bit less often (because this is the window from user container being alive and accepting connections -> returning 200), but no reason at all not to. Done.

Reply:
Also mesh — we specifically prohibit keep alives there.
If by upstream you mean ingress, kpa and activator probes.

====

Do we need `io.Copy` in defer though?

Reply:
THough on second thought: the probes to individual pods, can probably be keep-alive

Reply:
where we prohibit keep-alive we prohibit it by setting DisableKeepAlive: true on the client, there's never any harm copying the body to ioutil.Discard either way so the conn can be reused where keep alive is allowed (and in this specific case in the QP where we're doing it aggressively on a 25ms timer, we do want to reuse the connections). Honestly Im not sure if we need it in the defer or not, but it seemed logical to group it with the close. Happy to move it out of defer if you particularly prefer?

Reply:
Yep, copying won't make it worse. But I just wanted to highlight the general idea that we don't do it there.

Personal preference: not to defer what does not need to be deferred. But I am not bent on this here, since it also makes sense having them next to each other.